libsolidity/ast/Types.h

```diff
@@ -338,9 +384,13 @@ class IntegerConstantType: public Type
 
 	/// @returns the smallest integer type that can hold the value or an empty pointer if not possible.
 	std::shared_ptr<IntegerType const> integerType() const;
-
+	/// @returns the smallest fixed type that can hold the value or an empty pointer
+	std::shared_ptr<FixedPointType const> fixedPointType() const;
+	bigint denominator() const { return m_value.denominator(); }
+	
 private:
-	bigint m_value;
+	rational m_value;
+	short m_scalingFactor;
```

Comment:
So yes, the point where we have a problem is `fixed x = 1/3`, because it will never fit without truncating.
So the question is: What is the "intrinsic fixed point type" for a constant (what is returned by `fixedPointType()`) and which fixed point types do we allow implicit conversion to (explicit conversion is always fine)?

I would say the `canonical` fixed point type is the one with the smallest number of bits such that the number fits without losing information: I.e. we have to count the number of bits that can hold the numerator plus sign bit (should be the same code as in the integer case) and for the fractional bits, we should get that `denominator * 10**bits` is an integer. Then implicit conversions are ok if and only if the canonical fixed point type can be converted.